---
title: "Gene Expressions Corrleation-R Package"
author: "Felipe Barraza, Jon Blichar, Emmanuel Edu"
date: "`r Sys.Date()`"
output:
  pdf_document: default
  word_document: default
vignette: |
  %\VignetteIndexEntry{Vignette Title} %\VignetteEngine{knitr::rmarkdown} %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
## Overview
This developing vignette is for the `geneExpCor` Rpackage which is capable of receiving tissue data and outputting correlation values corresponding to matching genes. The objective of this software package is to identify putative endocrine interactions in a user-friendly way, and to compare inter-tissue communication discrepincies between human males and females. Research using this particular kind of analysis can be useful for understanding the functionality between varying organ pairs and to disease pathology to a genetic degree. For example, researchers found that men and women have different fat distribution patterns that impact their risk for metabolic and cardiovascular diseases. Women store excess adipose tissue in the gluteofemoral region while men store it around their abdominal region. This discrepancy in adipose tissue storage is a notable factor in understanding why men are at higher risk for cardiovascular disease. 

The goal of this work is to build upon analyzing inter-tissue endocrine interactions utilizing bioinformatics methodologies. Several functions are included that grant the user the capability to filter by variance percentage and by ligands. The 'geneExpCor' package will produce histograms and scatter plots corresponding to a biweight midcorrelation statistical analysis conducted on the inputted gene data sets. The built-in gene data sets, sourced from the GTEx consortium portal, are included to examine the capabilites of the functions within the package. 


## Dependencies
- `stats`: https://stat.ethz.ch/R-manual/R-devel/library/stats/html/stats-package.html 
- `dynamicTreeCut`: https://cran.r-project.org/web/packages/dynamicTreeCut/index.html 
- `fastcluster`: https://cran.r-project.org/web/packages/fastcluster/vignettes/fastcluster.pdf 
- `devtools`: https://cran.r-project.org/web/packages/devtools/index.html 
- `testthat`: https://testthat.r-lib.org/ Version 3.0 or above
- `WGCNA`: https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/ 
- `iemisc`: https://cran.rstudio.com/web/packages/iemisc/index.html 
- `dpylr`: https://www.tidyverse.org/ 
- `parallel`: https://stat.ethz.ch/R-manual/R-devel/library/parallel/doc/parallel.pdf 


## Installation

### 1. Installing From Github
After dependencies have been installed and loaded, installing `geneExpCor` from github requires the following line:  
```{r message = FALSE, warning= FALSE}
  library(devtools)
  devtools::install_github("Eziedu/Gene_Expression_Cors_JEF")
```
### 1.2 Addressing Installation Errors
If there is an error with loading `devtools`, it is most likely due to a version issue. The devtools package uses the `testthat`package, and in order for devtools to run appropriately the loaded `testhat` package needs to be of version 3.0 or higher.

If a simple update does not solve the problem, run the following line to update `testthat`:

`install.packages("rlang", dependencies = c("Depends", "Imports", "LinkingTo", "Suggests"))`

Follow the installation process as you see fit, afterwards the `testthat` package should update to version 3.0 or above. 

### 2. Loading Necessary Data
```{r message=FALSE, warning=FALSE}
library(stats) 
library(iemisc)
library(dynamicTreeCut) 
library(fastcluster) 
library(WGCNA)
library(dplyr)
library(parallel)
```
To properly use the following functions, the inputted data needs to be in a specific format. To allow for easy testing, the package comes with two preinstalled tissue datasets: `subqAdipose_tissue_data` and `stomach_tissue_data`. 

To properly access both datasets, we recommend attributing each set to a variable in the global enviornment. For example, run the following lines in the console: 
```{r}
library(geneExpCor)

tissue_1_data <- geneExpCor::subqAdipose_tissue_data
tissue_2_data <- geneExpCor::stomach_tissue_data

```

```{r include = FALSE}
gene_map <- geneExpCor::gencode_gene_map ###NEEDS TO BE REMOVED AT SOME POINT!!!
```

## Formatting Data 
To provide a working example, the `subqAdipose_tissue_data` and `stomach_tissue_data` data sets will be used as the primary tissue files when examining the vignette. 

### Filtering Data to Identify Matching Genes 

To begin analysis the `gene_names` function will be used first. `gene_names()` takes in two tissue data sets and attempts to match the genes within both tissues and returns two matrices for each tissue with the matching genes. 

```{r options(width = 300)}
gene_names(tissue_1_matrix = tissue_1_data, tissue_2_matrix = tissue_2_data, 
           ensembl_map = gene_map)
```
The outputted matrices, `final_tissue_1.matrix` and `final_tissue_2.materix`, are loaded in the global environment containing the matching donor and gene data present in both tissues. Examining the dimensions of the matrices can serve as a check to ensure that the outputs are appropriately filtered. The resulting matrices will have smaller dimensions that their original tissue sets. In the example, `final_tissue_1.matrix` has smaller dimensions compared to the original tissue data file, `tissue_1_data`.   

```{r}
# Checking Dimensions of original tissue files
dim(tissue_1_data) # 24665 BY 581
dim(final_tissue_1.matrix) # 22699 BY 244
```

The resulting matricies will have entries similar to the table below, where the rows represent the genes and columns are the samples. 
```{r}
final_tissue_1.matrix[1:3, 1:2]
```
After the raw data files have been formatted to relabel gene names and filtered for matching genes amongst both tissue sets, the `final_tissue.matrix` matricies are in the proper form and can be used for subsequent functions within the package. 

## Conducting Gene Analysis

### Filtering for Most Variable Genes
To filter tissue matricies according to variance, the `var_filter` function should be utilized. This function takes two tissue matrices and a variance percentage and outputs matrices of the most variable genes for each tissue set. This function is primarily utilized with tissue sets containing matching genes and donors: `final_tissue_1.matrix`. 
 
 This function will return two lists of the most variable genes based on the `variance_percentage` parameter. 
```{r}
var_filter(final_tissue_1.matrix, final_tissue_2.matrix, 0.3)
```
In the example above, the function is sifting through both gene data sets and filtering for the top 30 percent of most variable genes.


The outputted matricies `variable.tissue_1.genes` and `variable.tissue_2.genes` contain the most variable genes amongst both tissue sets and will therefore have smaller dimensions than the inputted tissue sets. The amount of donors, columns, remains the same, since the `var_filter()` is only filtering genes. 

Checking the dimensions is a brief method to ensure that the function is filtering the genes accordingly. For the var_filter(), the number of rows of the resulting matrices should be smaller than the inputted matrices. Additionally, both of the produced data sets will have equal dimensions.

In the ongoing example, the resulting tissue matrix,`variable.tissue_1.genes` has a smaller list of genes than the inputted matrix `final_tissue_1.matrix`. Check to ensure that this is the case. 
```{r}
#Check the Dimensions of the inputted matrices
dim(final_tissue_1.matrix) #22699 BY 244 

#Check the Dimensions of the outputted matricies 
dim(variable.tissue_1.genes) #6810 BY 244

dim(variable.tissue_2.genes) #6810 BY 244
```

The following is a selection from the `variable.tissue_1.genes` matrix acquired from the `var_filter` function using the data in the example.
```{r}
# View a section of the variance filtered matrix
variable.tissue_1.genes[3:5, 3:4]

```


### Conducting Biweight Midcorrelation Analysis
To conduct the biweight midcorrelation analysis on the tissue sets the `tissue_pair_gene_expression` function. 
This function takes in two tissue matrices and return two large matrices containing biweight midcorrelation coefficients and pvalues associated with both tissue sets. 

```{r message=FALSE, warning=FALSE}
tissue_pair_gene_expression(variable.tissue_1.genes, variable.tissue_2.genes)
```

The outputted variables, `gene_biweight_midcorrelation` and `pvalues_biweight` are used to identify the corrlelations of the genes in both tissues matricies and are utilized to develop a series of plots to visualize the data. 

## Generating Plots

To properly create visuals for the analysis the matrix the pvalue matrix and biweight correlations from the `tissue_pair_gene_expression` function. The following code block produces a histogram using the inverse log pvalue data.  
```{r}
# Create a matrix of inverse log base 10 p-values
logpvalues <- -log(pvalues_biweight)

# Create a histogram of inverse log base 10 p-values
hist(logpvalues,100)
```
 A histogram of the biweight correlations using the `gene_biweight_midcorrelation` matrix
```{r}
# Create a histogram of correlation coefficients
hist(gene_biweight_midcorrelation)
```
More customized plots can be developed by identifying certain pvalues that lie above a particular threshold. In this case, the cutoff considers log pvalues above 150. This value is arbitrary and can be altered to find a specific range of pvalues. 
```{r}
# Find the indices for the p-values greater than some cutoff value
cutoff.ind <- which(logpvalues >= 150, arr.ind = T)
```


The following code block generates a series of scatter plots according the size and values of the pvalue cutoff variable `cutoff.ind`. These scatter plots depict the pvalues between multiple gene pairs within the `final_tissue` matricies produced from the `gene_names` function.
```{r}
for (i in 1:(size(cutoff.ind,1) - 2)){
  # Find the indices for the individual gene pairs for each run through the for loop
  ind.variance <- which(pvalues_biweight == 
                          pvalues_biweight[cutoff.ind[i,1],cutoff.ind[i,2]], arr.ind = T)
  
  # Create variables containing the rownames and columnames to match gene names
  rows <- rownames(pvalues_biweight)
  columns <- colnames(pvalues_biweight)
  
  # Get the gene names for the highly correlated gene pair of interest
  gene1 = rows[ind.variance[1]]
  gene2 = columns[ind.variance[2]]
  
  # Find the indices of the highly correlated genes in our matrix of gene expressions
  gene1.ind = which(rownames(final_tissue_1.matrix) == gene1)
  gene2.ind = which(rownames(final_tissue_2.matrix) == gene2)
  # Plot the expressions of the two genes with high degree of correlation
  plot(as.numeric(final_tissue_1.matrix[gene1.ind,]),
       as.numeric(final_tissue_2.matrix[gene2.ind,]), 
       xlab="Expression in Subcutaneous Adipose Tissue", ylab="Expression in Stomach Tissue")
}
```

